Algorithmia/handler.py

```diff
@@ -0,0 +1,98 @@
+import sys
+import json
+import base64
+import traceback
+import six
+
+
+class Handler(object):
+
+    def __init__(self, apply_func, load_func=None):
+        """
+        Creates the handler object
+        :param apply_func: A required function that can have an arity of 1-2, depending on if loading occurs
+        :param load_func: An optional supplier function used if load time events are required, has an arity of 0.
+        """
+        self.FIFO_PATH = "/tmp/algoout"
+        self.apply_func = apply_func
+        if load_func:
+            self.load_func = load_func
+        else:
+            self.load_func = lambda: None
+
+    def load(self):
+        output = self.load_func()
+        print('PIPE_INIT_COMPLETE')
+        sys.stdout.flush()
+        return output
+
+    def format_data(self, request):
+        if request['content_type'] in ['text', 'json']:
+            data = request['data']
+        elif request['content_type'] == 'binary':
+            data = self.wrap_binary_data(request['data'])
+        else:
+            raise Exception("Invalid content_type: {}".format(request['content_type']))
+        return data
+
+    def is_binary(self, arg):
+        if six.PY3:
+            return isinstance(arg, base64.bytes_types)
+
+        return isinstance(arg, bytearray)
+
+    def wrap_binary_data(self, data):
+        if six.PY3:
+            return bytes(data)
+        else:
+            return bytearray(data)
+
+    def format_response(self, response):
+        if self.is_binary(response):
+            content_type = 'binary'
+            response = base64.b64encode(response)
+            if not isinstance(response, six.string_types):
+                response = str(response, 'utf-8')
+        elif isinstance(response, six.string_types) or isinstance(response, six.text_type):
+            content_type = 'text'
+        else:
+            content_type = 'json'
+        response_string = json.dumps({
+            'result': response,
+            'metadata': {
+                'content_type': content_type
+            }
+        })
+        return response_string
+
+    def write_to_pipe(self, data_string):
+        with open(self.FIFO_PATH, 'w') as f:
```

Comment:
Can we open the FIFO once in the constructor vs. opening a new file descriptor every time we write to it?

Reply:
there are some order of operations issues with this that only crop up during testing. I kept this as a strict refactor from the original pipe defined in langpacks, we could adjust this if we decide tho.

Reply:
That's a reasonable decision given the heritage. I'm not too hung up on it so I think it's fine to leave it as-is with that in mind.